libvips/foreign/cairo.c

```diff
@@ -127,21 +127,31 @@ vips__bgra2rgba(guint32 *restrict p, int n)
 	}
 }
 
+extern float vips_v2Y_16[65536];
+void vips_col_make_tables_RGB_16(void);
+
 /*
  * Convert from Cairo-style premultiplied RGBA128F to straight RGBA, for one row.
- *
+ * It also linearizes the pixel values.
+
  * Processes 'n' pixels in the 'p' buffer.
  * The data is assumed to be RGBA (R, G, B, A) 32-bit floats per pixel.
  */
 void
 vips__premultiplied_rgb1282scrgba(float *restrict p, int n)
 {
+	vips_col_make_tables_RGB_16();
 	for (int x = 0; x < n; x++) {
-		float r = p[0];
-		float g = p[1];
-		float b = p[2];
+		float r = VIPS_CLIP(0, p[0], 1);
+		float g = VIPS_CLIP(0, p[1], 1);
+		float b = VIPS_CLIP(0, p[2], 1);
 		float a = p[3];
 
+		// linearize the values with LUT
+		r = vips_v2Y_16[(int) (65535 * r)];
+		g = vips_v2Y_16[(int) (65535 * g)];
+		b = vips_v2Y_16[(int) (65535 * b)];
```

Comment:
nit: I was going to suggest doing `FCLIP` here (which is slower), but then I noticed we just want an int result. How about doing this instead?:
```suggestion
		// CLIP is much faster than FCLIP, and we want an int result
		int ri = VIPS_CLIP(0, (int) (p[0] * 65535), 65535);
		int gi = VIPS_CLIP(0, (int) (p[1] * 65535), 65535);
		int bi = VIPS_CLIP(0, (int) (p[2] * 65535), 65535);

		// linearize the values with LUT
		float r = vips_v2Y_16[ri];
		float g = vips_v2Y_16[gi];
		float b = vips_v2Y_16[bi];
		float a = p[3];
```